File: internal/poll/poll.go
package poll

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/wtgoku-create/popiartcli/internal/api"
	"github.com/wtgoku-create/popiartcli/internal/output"
	"github.com/wtgoku-create/popiartcli/internal/types"
)

var terminalStates = map[string]bool{
	"done":      true,
	"failed":    true,
	"cancelled": true,
}

func WaitForJob(ctx context.Context, client *api.Client, jobID string, interval time.Duration, maxPolls int) (*types.Job, error) {
	for pollIndex := 0; pollIndex < maxPolls; pollIndex++ {
		var job types.Job
		if err := client.GetJSON(ctx, "/jobs/"+jobID, nil, &job); err != nil {
			return nil, err
		}

		status := job.Status
		if terminalStates[status] {
			if status == "failed" {
				return nil, output.NewError("JOB_FAILED", messageFromJob(&job), map[string]any{
					"job_id": jobID,
					"status": status,
					"error":  job.Error,
				})
			}
			return &job, nil
		}

		fmt.Fprintf(os.Stderr, "\r⏳ %s - %s (%ds)   ", jobID, status, int(interval.Seconds())*pollIndex)
		time.Sleep(interval)
	}

	return nil, output.NewError("POLL_TIMEOUT", fmt.Sprintf("Job %s did not complete within the timeout", jobID), map[string]any{
		"job_id":          jobID,
		"timeout_seconds": int(interval.Seconds()) * maxPolls,
	})
}

func messageFromJob(job *types.Job) string {
	if job == nil || job.Error == nil {
		return "Job failed"
	}
	if job.Error.Message != "" {
		return job.Error.Message
	}
	return "Job failed"
}

File: src/lib/poll.js
// src/lib/poll.js
// 通用作业轮询器。当作业达到终止状态时退出。
// 终止状态：done（完成）、failed（失败）、cancelled（取消）
// 支持 --wait 标志：阻塞直到终止，在 stderr 上打印状态。
// 不带 --wait：立即返回 job_id。

import { get } from './client.js';
import { out, err } from './output.js';

const TERMINAL = new Set(['done', 'failed', 'cancelled']);
const DEFAULT_INTERVAL_MS = 2000;
const DEFAULT_MAX_POLLS = 300; // 在2秒间隔下为10分钟

export async function pollJob(jobId, {
  wait = false,
  intervalMs = DEFAULT_INTERVAL_MS,
  maxPolls = DEFAULT_MAX_POLLS,
  plain = false,
} = {}) {
  if (!wait) {
    return out({ job_id: jobId, status: 'pending', polling: false }, { plain });
  }

  let polls = 0;
  while (polls < maxPolls) {
    const job = await get(`/jobs/${jobId}`);
    const status = job.status;

    if (TERMINAL.has(status)) {
      if (status === 'failed') {
        err('JOB_FAILED', job.error?.message ?? 'Job failed', {
          job_id: jobId,
          status,
          error: job.error,
        });
      }
      return out(job, { plain });
    }

    // 在 stderr 上打印进度提示 (避免污染 stdout 中的 JSON)
    process.stderr.write(`\r⏳ ${jobId} — ${status} (${polls * intervalMs / 1000}s)   `);
    await sleep(intervalMs);
    polls++;
  }

  err('POLL_TIMEOUT', `Job ${jobId} did not complete within the timeout`, {
    job_id: jobId,
    polls,
    timeout_ms: polls * intervalMs,
  });
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
